src/lib/y2storage/btrfs_subvolume.rb

```diff
@@ -187,69 +207,57 @@ def exclusive_limit=(limit)
-    #
-    # @return [Boolean] true if the subvolume is shadowed
-    def shadowed?
-      !shadowers.empty?
-    end
+    protected
 
-    # Returns the devices that shadow the subvolume
+    # Whether the subvolume requires a default mount point
     #
-    # It prevents to return the subvolume itself or its filesystem as shadower.
-    #
-    # @return [Array<Mountable>] shadowers
-    def shadowers
-      shadowers = BtrfsSubvolume.shadowers(devicegraph, mount_path)
-      shadowers.reject { |s| s.sid == sid || s.sid == btrfs.sid }
-    end
-
-    # Checks whether a mount path is shadowing another mount path
+    # Only subvolumes for root require a default mount point. This is necessary to correctly mount the
+    # subvolumes when snapshoting is used. Also note that a subvolume does not require a mount point when
+    # any of its ancestors already has a mount point or when the subvolume is a snapshot.
     #
-    # @note The existence of devices with that mount paths is not checked.
+    # Example:
     #
-    # @param mount_path [String]
-    # @param other_mount_path [String]
+    # top-level                       -> false
+    # |-- @                           -> false
+    #    |-- var                      -> true
+    #    |  |-- log                   -> false
+    #    |-- .snapshots               -> false
+    #        |--.snapshots/1/snapshot -> false
+    # |-- foo                         -> true
     #
-    # @return [Boolean] true if other_mount_path is shadowed by mount_path
-    def self.shadowing?(mount_path, other_mount_path)
-      return false if mount_path.nil? || other_mount_path.nil?
-      return false if mount_path.empty? || other_mount_path.empty?
-
-      # Just checking with start_with? is not sufficient:
-      # "/bootinger/schlonz".start_with?("/boot") -> true
-      # So append "/" to make sure only complete subpaths are compared:
-      # "/bootinger/schlonz/".start_with?("/boot/") -> false
-      # "/boot/schlonz/".start_with?("/boot/") -> true
-      check_path = "#{other_mount_path}/"
-      check_path.start_with?("#{mount_path}/")
+    # @return [Boolean]
+    def require_default_mount_point?
+      return false unless filesystem.root?
+      return false if top_level? || default_btrfs_subvolume? || snapshot?
+
+      parent_subvolume.top_level? || parent_subvolume.prefix?
     end
 
-    # Checks whether a mount path is currently shadowed by any other mount path
-    #
-    # @param devicegraph [Devicegraph]
-    # @param mount_path [String] mount point to check
+    # Whether the subvolume is for snapshots
     #
     # @return [Boolean]
-    def self.shadowed?(devicegraph, mount_path)
-      !shadowers(devicegraph, mount_path).empty?
+    def snapshot?
+      path.match?(/^.snapshots/)
     end
 
-    # Returns the current shadowers for a specific mount point
+    # Whether the subvolume is the used as prefix (typically @)
```

Comment:
Maybe we should mention that this is only trustworthy if we have volume specifications (ie. during installation).

Reply:
Actually not. The control file is also accessible in the running system.

Reply:
But is trustworthy in that case? Does it really represent the product/role that was installed and all that?

Reply:
I am not sure, but I think that the control file stored in the installed system represents the final version. Anyway, as we have commented on IRC, in case of AutoYaST we should take into account the AutoYaST profile to decide whether a subvolume is the prefix. I will improve that as part of https://github.com/yast/yast-storage-ng/pull/1168.